packages/patrol_cli/lib/src/features/test/android_test_backend.dart

```diff
@@ -60,74 +92,126 @@ class AndroidAppOptions {
   }
 }
 
-class AndroidTestBackend {
+/// Provides functionality to build, install, run, and uninstall Android apps.
+///
+/// This class must be stateless.
+class AndroidTestBackend extends TestBackend {
   AndroidTestBackend({
+    required Adb adb,
     required ProcessManager processManager,
     required Platform platform,
     required FileSystem fs,
     required DisposeScope parentDisposeScope,
     required Logger logger,
-  })  : _processManager = processManager,
+  })  : _adb = adb,
+        _processManager = processManager,
         _fs = fs,
         _platform = platform,
         _disposeScope = DisposeScope(),
         _logger = logger {
     _disposeScope.disposedBy(parentDisposeScope);
   }
 
+  final Adb _adb;
   final ProcessManager _processManager;
   final Platform _platform;
   final FileSystem _fs;
   final DisposeScope _disposeScope;
   final Logger _logger;
 
-  Future<void> run({
-    required Device device,
-    required AndroidAppOptions options,
-  }) async {
-    final targetName = basename(options.target);
-    final task = _logger
-        .task('Building apk for $targetName and running it on ${device.id}');
-
-    final process = await _processManager.start(
-      options.toGradleInvocation(_platform),
-      runInShell: true,
-      environment: {
-        'ANDROID_SERIAL': device.id,
-      },
-      workingDirectory: _fs.currentDirectory.childDirectory('android').path,
-    );
-
-    process
-        .listenStdOut((line) => _logger.detail('\t: $line'))
-        .disposedBy(_disposeScope);
-    process
-        .listenStdErr((line) => _logger.err('\t$line'))
-        .disposedBy(_disposeScope);
+  @override
+  Future<void> build(AndroidAppOptions options, Device device) async {
+    await _disposeScope.run((scope) async {
+      final subject = options.desc;
+      final task = _logger.task('Building $subject');
+
+      Process process;
```

Comment:
`late final`? 👀 this shouldn't be a var

Reply:
This `process` variable is assigned twice, to different processes.